tests/test_schema.py

```diff
@@ -1745,14 +1745,22 @@ class Model(BaseModel):
         a: Literal[1]
         b: Literal['a']
         c: Literal['a', 1]
+        d: Literal['a', Literal['b'], 1, 2]
 
     assert Model.schema() == {
         'properties': {
-            'a': {'title': 'A', 'type': 'integer', 'const': 1},
-            'b': {'title': 'B', 'type': 'string', 'const': 'a'},
-            'c': {'anyOf': [{'type': 'string', 'const': 'a'}, {'type': 'integer', 'const': 1}], 'title': 'C'},
+            'a': {'title': 'A', 'type': 'integer', 'enum': [1]},
```

Comment:
Worth comparing with the first commit https://github.com/samuelcolvin/pydantic/pull/2348/commits/3a14c7219bb9b00d20f7c4fbe44ba7b404875100 😉